### PR TITLE
[resharding] Force creation of snapshot for resharding

### PR DESCRIPTION
**Pull request by shreyan-gupta**
This PR introduces a `force_snapshot` bool which overrides the snapshot config for creating snapshot. This is useful for resharding where we are required to use a snapshot for creating child tries.

### CODE REVIEW

File: integration-tests/src/tests/client/resharding.rs [hunk 1]
```diff
@@ -190,10 +191,14 @@ impl TestReshardingEnv {
             genesis_protocol_version,
         );
         let chain_genesis = ChainGenesis::new(&genesis);
-        let env = TestEnv::builder(chain_genesis)
+        let builder = if state_snapshot_enabled {
```
Comment (shreyan-gupta) [inline integration-tests/src/tests/client/resharding.rs:194]: This is super cool as now we can run the resharding tests without enabling snapshot. For testing purposes, I have tests for both when the snapshot config is enabled and disabled.

File: debug_scripts/Pipfile [hunk 2]
```diff
@@ -5,7 +5,7 @@ name = "pypi"
 
 [packages]
 boto3 = "==1.21.34"
-requests = "==2.27.1"
+requests = "==2.31.0"
```
Comment (shreyan-gupta) [inline debug_scripts/Pipfile:8]: `debug_scripts/Pipfile`, `debug_scripts/Pipfile.lock` and `tools/debug-ui/package-lock.json` are not my changes. They may disappear once I sync with master. Meanwhile please ignore these files.

**Comment by nikurt** [comment]
I find the proposed interface confusing.
Having both `enabled` and `force_enabled` leads to combinations which are hard to maintain and reason about.

Suggestion:
Change the interface such that:
If `MakeSnapshotRequest` is received, then don't do any checks and simply proceed to making the snapshot.
Change `make_state_snapshot()` to unconditionally make a snapshot.

`enabled` is confusing if it can be overridden by `force_snapshot`, maybe rename it to `snapshot_every_epoch`. Or even better, rename it to `snapshot_schedule` and remove the option of `snapshot_every_n_epochs`.

**Review by wacban — COMMENTED** [review]
LGTM but nikurt's suggestion is great too, please discuss it with him. The snapshot_schedule in config can be an enum with options such as "EveryEpoch, "EveryEpochAndNBlocks", "WhenNeeded"/ "PreResharding".

File: chain/client/src/test_utils/test_env_builder.rs [hunk 3]
```diff
@@ -360,10 +365,15 @@ impl TestEnvBuilder {
 
     /// Internal impl to make sure runtimes are initialized.
     fn ensure_runtimes(self) -> Self {
+        let state_snapshot_enabled = self.state_snapshot_enabled;
```
Comment (walnut-the-cat) [inline chain/client/src/test_utils/test_env_builder.rs:368]: just out of curiosity, why do we need to do this allocation?

Comment (shreyan-gupta) [reply chain/client/src/test_utils/test_env_builder.rs]: `self` gets "consumed" in the call to `ensure_epoch_managers()` and can not be used later to access any variable. We can only access `ret` but not it's private members like `state_snapshot_enabled`

**Pull request closed without merging**